db/view/view.cc

```diff
@@ -1232,6 +1250,59 @@ void view_updates::generate_update(
     }
 }
 
+std::optional<partition_key> view_updates::construct_view_partition_key_from_base(const partition_key& base_pk)
+{
+    // We check that the view partition key is a permutation of the
+    // base partition key. If so, we can construct the corresponding
+    // view partition key from the base key and apply an optimized
+    // partition level update. Otherwise, we return std::nullopt.
+
+    if (!_view_info.is_partition_key_permutation_of_base_partition_key()) {
+        return std::nullopt;
+    }
+
+    auto base_exploded_pk = base_pk.explode();
+    std::vector<bytes> view_exploded_pk(_view->partition_key_size());
+
+    // Construct the view partition key by finding each component
+    // in the base partition key.
+    for (const column_definition& view_cdef : _view->partition_key_columns()) {
+        const column_definition* base_cdef = _base->get_column_definition(view_cdef.name());
+        if (base_cdef && base_cdef->is_partition_key()) {
+            view_exploded_pk[view_cdef.id] = base_exploded_pk[base_cdef->id];
+        } else {
+            // This shouldn't happen because we already checked that all
+            // the view partition key columns appear in the base partition key.
+            // But if for some unexpected reason we get here, we can return nullopt
+            // and skip the optimization.
+            return std::nullopt;
```

Comment:
I added a rate-limited warning

Reply:
Since just a few lines above we check is_partition_key_permutation_of_base_partition_key() which is supposed to guarantee we don't have this problem, I wonder if it's worth it to do anything besides an on_interal_error()?

Reply:
it would be good to catch it and know if it happens, so it can be fixed quickly
on the other hand, if it does, I wouldn't want it to crash the node or fail the view update

Reply:
on_internal_error() doesn't crash the node (only in testing), it's a normal exception.
Anyway, my point was that just a few lines above, you guarantee that this case can't happen. So while it's good to be safe (the saying goes "a programmer is someone who looks both ways when crossing a one-way street"), I don't think it's worth 10 lines of code of logging, rate limiting, etc. Unless I'm missing something and the check a few lines above doesn't exclude this case?

Reply:
I changed it to on_internal_error